src/pog.gleam

```diff
@@ -911,3 +947,81 @@ fn seconds_decoder() -> decode.Decoder(#(Int, Int)) {
 pub fn numeric_decoder() -> decode.Decoder(Float) {
   decode.one_of(decode.float, [decode.int |> decode.map(int.to_float)])
 }
+
+pub type Range(t) {
+  Range(lower: Bound(t), upper: Bound(t))
+  Empty
+}
+
+pub type Bound(t) {
+  Bound(value: t, inclusivity: Inclusivity)
+  Unbound
+}
+
+pub type Inclusivity {
+  Inclusive
+  Exclusive
+}
+
+pub fn range_decoder(
+  value_decoder: decode.Decoder(t),
+) -> decode.Decoder(Range(t)) {
+  let range_decoder = {
+    use lower_inclusivity <- decode.subfield([1, 0], inclusivity_decoder())
+    use upper_inclusivity <- decode.subfield([1, 1], inclusivity_decoder())
+    use lower_bound <- decode.subfield(
+      [0, 0],
+      bound_decoder(value_decoder, lower_inclusivity),
+    )
+    use upper_bound <- decode.subfield(
+      [0, 1],
+      bound_decoder(value_decoder, upper_inclusivity),
+    )
+    decode.success(Range(lower_bound, upper_bound))
+  }
+
+  let empty_decoder = {
+    use decoded_atom <- decode.then(atom.decoder())
+    case decoded_atom == atom.create("empty") {
+      True -> decode.success(Empty)
+      False -> decode.failure(Empty, "empty")
```

Comment:
What's `empty`? Shouldn't this be `Range`?

Reply:
As far as I understand, this is a _sub decoder_ designed to decode only a single `empty` Erlang atom. Its error is always suppressed by the usage of `decode.one_of`.
I've changed the expected error message to be clearer:
```rust
False -> decode.failure(Empty, "`empty` atom")
```

I've also changed the main/parent decoder to collapse errors into something more readable:
```rust
decode.one_of(range_decoder, [empty_decoder])
|> decode.collapse_errors("`#(#(t, t), #(bool, bool))` tuple or `empty` atom")
```

Is it better now or could you please explain what I'm missing here?